compiler/src/dotty/tools/dotc/transform/InstrumentCoverage.scala

```diff
@@ -55,186 +57,333 @@ class InstrumentCoverage extends MacroTransform with IdentityDenotTransformer:
       val pos = tree.sourcePos.withSpan(tree.span.withEnd(friendlyEnd)) // user-friendly span
-      // ensure that the body is always instrumented by inserting a call to Invoker.invoked at its beginning
-      val instrumentedBody = instrument(transform(tree.body), pos, false)
-      cpy.CaseDef(tree)(tree.pat, transform(tree.guard), instrumentedBody)
 
-    /** Records information about a new coverable statement. Generates a unique id for it.
-      * @return the statement's id
-      */
-    private def recordStatement(tree: Tree, pos: SourcePosition, branch: Boolean)(using ctx: Context): Int =
-      val id = statementId
-      statementId += 1
-      val statement = Statement(
-        source = ctx.source.file.name,
-        location = Location(tree),
-        id = id,
-        start = pos.start,
-        end = pos.end,
-        line = pos.line,
-        desc = tree.source.content.slice(pos.start, pos.end).mkString,
-        symbolName = tree.symbol.name.toSimpleName.toString,
-        treeName = tree.getClass.getSimpleName.nn,
-        branch
-      )
-      coverage.addStatement(statement)
-      id
-
-    private inline def syntheticSpan(pos: SourcePosition): Span = pos.span.toSynthetic
-
-    /** Shortcut for instrument(tree, tree.sourcePos, branch) */
-    private inline def instrument(tree: Tree, branch: Boolean = false)(using Context): Tree =
-      instrument(tree, tree.sourcePos, branch)
+      // recursively transform the guard, but keep the pat
+      val transformedGuard = transform(guard)
 
-    /** Instruments a statement, if it has a position. */
-    private def instrument(tree: Tree, pos: SourcePosition, branch: Boolean)(using Context): Tree =
-      if pos.exists && !pos.span.isZeroExtent then
-        val statementId = recordStatement(tree, pos, branch)
-        insertInvokeCall(tree, pos, statementId)
-      else
-        tree
-
-    /** Instruments the body of a DefDef. Handles corner cases. */
+      // ensure that the body is always instrumented by inserting a call to Invoker.invoked at its beginning
+      val coverageCall = createInvokeCall(tree.body, pos)
+      val transformedBody = transform(tree.body)
+      val instrumentedBody = InstrumentedParts.singleExprTree(coverageCall, transformedBody)
+
+      cpy.CaseDef(tree)(pat, transformedGuard, instrumentedBody)
+
+    /** Transforms the parents of a Template. */
+    private def transformTemplateParents(parents: List[Tree])(using Context): List[Tree] =
+      def transformParent(parent: Tree): Tree = parent match
+        case tree: Apply =>
+          // only instrument the args, not the constructor call
+          cpy.Apply(tree)(tree.fun, tree.args.mapConserve(transform))
+        case tree: TypeApply =>
+          // args are types, instrument the fun with transformParent
+          cpy.TypeApply(tree)(transformParent(tree.fun), tree.args)
+        case other =>
+          transform(other)
+
+      parents.map(transformParent)
```

Comment:
I assume most of the time there won't be anything to transform so we might as well do:
```suggestion
      parents.mapConserve(transformParent)
```